### ISSUE TITLE
[Questions and issues] GPflow 1.0 Release Candidate

### ISSUE THREAD

**Issue by awav** [GPflow/GPflow#519]
Hello everyone,

TL;DR: We have redesigned GPflow: modified some method protocols, added new capabilities and features. We'd like to get your feedback and collect your requests and cases, especially those which you think became harder to implement or which you simply can't make work with new GPflow.

Please ask any questions, even simpliest ones like _how to create a parameter_. Of course the thread can end up in chaos without elementary rules, here they are: _question, request or idea must be comprehensible_, to achieve this provide a short example what you are doing and what you want to get with new GPflow or which feature you think is important to have. If you found a bug or you think GPflow has unclear behaviour, an error message, accompany your thorough description with a code example -  _gist, notebook, code snippet_ or whatever. Please, also include _SHA of GPflow-1.0-RC_, _OS_, _Python_ and _TensorFlow_ versions which you used in your experiments. Reviewer should be able to copy-paste and run your code on local machine.

We look forward hearing from you!

P.S. If you have not checked out new GPflow, please do it: [gpflow1.0-rc](https://github.com/GPflow/GPflow/tree/GPflow-1.0-RC).

**Comment by javdrher** [comment]
This week I spent some time reworking GPflowOpt to use the new RC. I haven't gotten far yet but already bumped into quite some issues. Some of them were some internal issues of GPflow which @awav fixed, thanks a lot for that.

I'm excited about all the new possibilities the RC offers. However, right now it seems this comes at a considerable cost as some functionality disappeared: 
- The old GPflow would manage the graph and session for you. The RC requires the user to do that.
- Because of this, the user needs to be very well aware about the internals of GPflow which will require a lot of documentation and result in a steep learning curve. You need to be careful: e.g., calling compile() again can undo changed you made to values. In scenarios beyond setting up GPR for a regression this can cause bugs which are difficult to track down.

To illustrate, I provide some runnable examples related to use cases in GPflowOpt. I found workarounds for most of them, but they illustrate behavior which I believe is complicated for a user unfamiliar with the entire GPflow codebase which grew considerably in size. The examples were made with python 3.5.2, SHA 0400eb8a2a921d0dce24b7fb92e55bc458f27301 and tensorflow 1.3.0

To avoid a lot of code I implemented the code concepts instead of the actual implementations in GPflowOpt. I'll describe the use case though to motivate why its happening. 

GPflowOpt has a model wrapper class for implementing data scaling (e.g output normalization). In the constructor transform classes are constructed, and the data of a model is changed. When calling predict_f the scaler scales the input and the returned mu and var of the model, this is then integrated in the acquisition class. Something unexpected happens when assigning the model in the scaler:
https://gist.github.com/javdrher/a277f4c88a68454d3fcd6aeaa8c974a6
What about first assigning, then scaling? This delays the problem:
https://gist.github.com/javdrher/8465cbf77b0927f37a69e748d57b6958
Ultimately I found a fix. Note the clear in the constructor of Scaler, its crucial:
https://gist.github.com/javdrher/b986157c53370e1e1c453ad82cee8969

I believe the same behavior is also problematic in case of the VGP. Changing X/Y in VGP requires a recompile as the shape of the q_mu and q_sqrt should change accordingly. However, when calling `compile()` I expected X and Y to revert to the values before the update, as well as the parameters. However, when I created an MWE another issue popped up (I noticed that in compile, no lower triangular transform is used for q_sqrt, this might actually be incorrect?):
https://gist.github.com/javdrher/1af0b8349962687488b3cfe02556cd54
When specifying a new graph for the second compile, it works. Using only a clear does not:
https://gist.github.com/javdrher/c2b7c4a774119eb8606f1540ab6ba9e8

High-level, the control loop of GPflowOpt looks as follows (pseudocode) which is also a problem
```python
for i in range(no_iterations)
   model.optimize()
   newX = opt.maximize(acquisition)
   newY = objective(newX)
   model.X = np.vstack((model.X, newX))
   model.Y = np.vstack((model.Y, newY))
   user_callback(model) # allow user to change priors, configure initial starting point etc.
```
This is no longer possible because setting X and Y, as well as the user_callback may require a `compile()`. In the current master of GPflow, the recompilation feature would take care of instantiating a new graph, if required as this was governed by GPflow. With the RC I think we need to call compile() every iteration as we don't know if a recompilation is required or not. This would undo the update of model.X/Y so we would need to create a new graph as well. This means we would implement +/- the compilation mechanism of the current GPflow, but loose the ability to recompile only if it is required.

**Comment by markvdw** [comment]
I love it so far, it's much more customisable.

I'm looking for a replacement to `model.get_parameter_dict()` and `model.set_parameter_dict()`. Generally I think it would be nice to be able to get a data structure out of a model, which specifies the state of all the parameters. I can see this being useful for two main reasons:
- Temporarily storing model states in Python objects
- Simply restoring all the variables from a Python data structure.

I believe this would also be useful in performing MCMC, as predicting would require sequentially setting the model state to a bunch of different parameter settings.

What I specifically want to use it for, is to initialise a new model based on a modified copy of an old model. So something like this would be ideal:
```
vfe = gpflow.models.SGPR(X, Y, gpflow.kernels.RBF(1), Zinit)
param_dict = get_parameter_dict(init_params)
vfe.compile()
gpflow.train.ScipyOptimizer().minimize(vfe)

param_dict['model.Z'] = np.vstack((param_dict['model.Z'], X[None, np.random.randint(10), :]))
upper = gp_upper.SGPU(X, Y, gpflow.kernels.RBF(1), Z=param_dict['model.Z'])
upper.set_parameter_dict(init_params)
```

In this example, I want to initialise a different model `SGPU` which has all the same fields as `SGPR` to a slightly modified version of `SGPR`.

A slight added complication is that the shape of `model.Z` changes. It requires ensuring that the corresponding `Param` object has the correct size, which here is ensured in the constructor. However it would be nice if this was handled automatically as well, even though it wasn't in the past.

**Comment by dustinvtran** [comment]
Super excited for this. For the 1.0rc release, are there any plans regarding changes to backends? For example, replacing `densities.py`, `transforms.py`, and the sampling in`priors.py` with `tf.contrib.distributions`' distributions and bijectors (#332)? A good number of them are already in [core TF](https://www.tensorflow.org/versions/master/api_docs/python/tf/distributions) over [TF contrib](https://www.tensorflow.org/versions/master/api_docs/python/tf/contrib/distributions). This would also immediately open the door to multivariate priors (#450).

**Comment by awav** [comment]
Hello everyone,

@javdrher, I do believe that latest commits address most of the issues which you listed here, I successfully ran your code, see comments to the gists.
Btw, [vgp_broken_1](https://gist.github.com/javdrher/1af0b8349962687488b3cfe02556cd54) and [vgp_borken_2](https://gist.github.com/javdrher/c2b7c4a774119eb8606f1540ab6ba9e8) are different gists with same content. I checked VGP, this is definitely a bug. In fact, VGP must be rewritten according to new design, so that it will no longer required to re-compile VGP each time, when you change X and Y.
>  This means we would implement +/- the compilation mechanism of the current GPflow, but loose the ability to recompile only if it is required.
That's fair, I have not found solution for re-compilation feature. So far, it is a trade-off between introducing a bunch of features and not calling compile method. Btw, in terms of performance - rebuilding the graph is not so bad comparing to old GPflow parameters state handling ;)

@markvdw, I can be wrong, but even in old GPflow it is not trivial task. As far as I remember model state is built using sorted by names parameter list which is fitted into one variable. Here, main issue comes - what are those slightly different models, do they have same names for parameters, same number of parameters? They have different variable layouts in both cases. In current implementation it is even harder to find out either these models are "slightly" different or "mostly" similar. At first glance it is error prone approach. You have a full control over your model, and there is a simple solution for copying parameters from one model to the same model: 
`[param.assign(value) for param, value in list(zip(m1.trainable_parameters, m2.read_trainables()))]`. The rest is only your imagination, for e.g. you can create custom copy constructor for your model.

@dustinvtran, `1.0rc` replaces "engine" at GPflow with minimal change in framework's feature space, but it also boosts GPflow capabilities via superior TensorFlow exploitation. I've been restraining myself from changing the code and introducing more and more features. High priority is to finish core changes at GPflow and then we will proceed with distinct features. 
It is good to have this request anyway. We need to keep track of features which people think are important. Btw, I was about to replace `densities.py`, but put this thought and code aside :)

**Comment by alexggmatthews** [comment]
hi @awav we will only be using distributions from core not from contrib as per #332 . I suggest you continue to `restrain yourself' as we would not except a pull request that went beyond the scope of the changes already discussed. after this rather large PR we will be back to the standard way of working i.e one PR one feature/fix which helps the proper review process.

**Comment by markvdw** [comment]
@awav: Ah yes, this is a super neat solution. Essentially `model.get_parameter_dict()` can be written as:
```
param_dict = {p[0].full_name.replace('SGPR', 'SGPU'): p[1] for p in zip(vfe.trainable_parameters, vfe.read_trainables())}
```

Eventually I settled on changing `model.set_parameter_dict()` to:
```
assign_params = {p.full_name: p for p in upper.trainable_parameters}
[assign_params[k].assign(param_dict[k]) for k in param_dict]
```

This ensures that the parameters do not need to be in the same order in both models for the code to work. I would suggest that `model.set_parameter_dict()` in the form above would be included in the new code. I know it's only a two liner, but it takes a bit of thought and I think it's common. I'm happy to put this in a PR at a convenient time.

**Comment by awav** [comment]
@alexggmatthews,

> we would not except a pull request that went beyond the scope of the changes already discussed

That's right. I do not want to inflate this heavy PR neither.

@markvdw, how to check that they have the same contract? There is no guarantee that full names are equal for these models, they can be indexed with prefix '[INDEX]/[FULL NAME]'. Anyway, we could overcome this too.

**Comment by markvdw** [comment]
I agree, I don't think that check should be necessary. It should be up to the user to make sure that the models have the same contract. In my code, if there was a key in `param_dict` that was not in the target model, this would result in a `KeyError`. Otherwise all parameters in `param_dict` would be set in the target model.

**Comment by markvdw** [comment]
I also just thought about another use-case. @jameshensman mentioned that he thought it was weird that we're doing `gpflow.train.ScipyOptimizer().minimize(vfe)`. Perhaps we should be explicit about _which_ tensor of the model we are optimising? Something like `gpflow.train.ScipyOptimizer().minimize(vfe.negative_likelihood_tensor)`?

I haven't looked at the internals, so I don't know exactly how feasible this is. However, I do think this is useful as it also allows one model to define multiple optimisation objectives.

**Comment by awav** [comment]
@markvdw, `optimizer` must have an information about trainable variables, otherwise it will not know how to initialize correctly optimizer's minimization step. Passing GPflow model to optimizer always means you will optimize `model.objective`. I considered adding an ability to pass list of tensors and GPflow models to optimizer, but at that time I felt it is unnecessary and there is no hurry with it.

**Comment by awayt039** [comment]
Hello, I am relatively new to python and newer to GP modeling. Python and GPRs are by no means standard in my field but I am interested in learning more. As I was considering packages to use I came across this pymc3 example which models, what I would call, random effects http://docs.pymc.io/notebooks/GLM-hierarchical.html. I was curious if gpflow or gpy have a similar way to model random effects. I tried googling around but it is slightly confusing as different disciplines you different terminology to describe random effects/hierarchical model/multilevel model. I have seen some GPy examples applied to gene studies that use "hierarchical" models. Cool. However, I have seen the term hierarchical models applied to all sorts of different models that are not what I would call "random effects." Would anyone be able to provide some clarity to this novice? Thanks!

**Comment by dtpc** [comment]
Hi, there is a bug in that a `Parameterized` object can not be copied. The following fails
``` python
import gpflow as gp
import copy
p = gp.params.Parameterized(name='param')
copy.copy(p)
```
with `AttributeError: 'Parameterized' object has no attribute '_parent'`

A simple fix is swap the order in the if statement below:
```
diff --git a/gpflow/params/parameterized.py b/gpflow/params/parameterized.py
index de1e323..351d59f 100644
--- a/gpflow/params/parameterized.py
+++ b/gpflow/params/parameterized.py
@@ -242,7 +242,7 @@ class Parameterized(Node):
 
     def __getattribute__(self, name):
         attr = misc.get_attribute(self, name)
-        if TensorConverter.tensor_mode(self) and isinstance(attr, Parameter):
+        if isinstance(attr, Parameter) and TensorConverter.tensor_mode(self):
             return Parameterized._tensor_mode_parameter(attr)
         return attr
```

**Comment by nknudde** [comment]
Hi all,

I'm trying to use variables inside a model that are not Params. Unfortunately the following snippet does not work due to initialization issues. Is there a quick workaround for this issue available?

```python
import gpflow
import tensorflow as tf

class Demo(gpflow.models.Model):
    def __init__(self):
        super().__init__()
        self.a = gpflow.Param(10.0)

    @gpflow.params_as_tensors
    def _build_likelihood(self):

        # get and initialize a variable
        b = tf.get_variable('foo', shape=(), dtype=gpflow.settings.tf_float)

        return self.a ** 2 + b ** 2


def main():
    tf.reset_default_graph()
    m = Demo()
    m.compile()
    m.session.run(tf.global_variables_initializer())
    print(m.compute_log_likelihood())
    return m

if __name__ == "__main__":
    m = main()
```

**Comment by javdrher** [comment]
Small thing I noticed, you can instantiate the abstract classes in `gpflow.core`

```python
>>> import gpflow
>>> a = gpflow.core.base.IPrior()
>>> type(a)
<class 'gpflow.core.base.IPrior'>
```
adding `__metaclass__` in the object seems to be the python 2.7 way of making abstract base classes, in python 3 it is done as follows:
```python
class IPrior(metaclass=ABCMeta)
     ....
```

**Comment by awav** [comment]
I apologize for not answering you guys, I was extremely overloaded this month. And also for some reason, I did not receive notifications :)

1. @dtpc, yeah, that's a bit tricky. Btw, tensorflow variable cannot be copied. `copy.copy` creates another python object, but the variables stays the same.

2. @nknudde
```
class Demo(gpflow.models.Model):
    def __init__(self):
        super().__init__()
        self.a = gpflow.Param(10.0)
        self.b = tf.get_variable('foo', shape=(), dtype=gpflow.settings.tf_float)
    
    def initializables(self):
           return super().initializables + [self.b]

    @gpflow.params_as_tensors
    def _build_likelihood(self):
        return self.a ** 2 + self.b
```

3. @javdrher, already resolved as we shifted to python3. It was ugly `compatibility` hack, which didn't actually work :)

**Comment by awav** [comment]
GPflow 1.0 released. Closing this issue.

**Comment by RodolfoFerro** [comment]
Hey, congrats for the release! 🎉 

@awav Are new examples in the plan for updates in documentation?
I was wondering because I was using a previous version and now that I upgraded to RC-1.0 I get errors like `GPflowError: Model is not built.` and when I try to build it, I get other errors.
 
Also thanks @markvdw, I'm working on updating models' parameters and your `param_dict` proposal seems to be the solution for what I need.
